colibri/src/blockchain/node_inquirer.rs

```diff
@@ -0,0 +1,141 @@
+            }
+        };
+
+        let mut connected_count = 0;
+        for node in rpc_nodes {
+            match self.connect_to_node(&node).await {
+                Ok(_) => {
+                    debug!("Successfully connected to {} node for {}", node.name, self.blockchain.as_str());
+                    connected_count += 1;
+                },
+                Err(e) => {
+                    error!("Failed to connect to {} node for {}: {}", node.name, self.blockchain.as_str(), e);
+                }
+            }
+        }
+
+        connected_count
+    }
+
+    async fn connect_to_node(&self, node: &RpcNode) -> Result<(), String> {
+        let provider =  ProviderBuilder::new()
+            .on_http(node.endpoint.parse().unwrap());
+
+        let mut mapping = self.provider_mapping.write().await;
+        mapping.insert(
+            node.clone(),
+            Box::new(DynProvider::new(provider)),
+        );
+        Ok(())
+    }
+
+    pub async fn connected_to_any_provider(&self) -> bool {
+        !self.provider_mapping.read().await.is_empty()
+    }
+}
+
+// A simple manager that stores EVM node inquirers for different chains
+pub struct EvmInquirerManager {
+    inquirers: RwLock<HashMap<SupportedBlockchain, Arc<EvmNodeInquirer>>>,
+    globaldb: Arc<GlobalDB>,
+}
+
+impl EvmInquirerManager {
+    pub fn new(globaldb: Arc<GlobalDB>) -> Self {
+        Self {
+            inquirers: RwLock::new(HashMap::new()),
+            globaldb,
+        }
+    }
+
+    #[allow(dead_code)]
+    /// Initialize inquirers for all supported blockchains
+    pub async fn init_all_inquirers(&self) {
+        debug!("Starting initialization of EVM inquirers for all supported blockchains");
+        for blockchain in [
+            SupportedBlockchain::Ethereum,
+            SupportedBlockchain::Optimism,
+            SupportedBlockchain::PolygonPos,
+            SupportedBlockchain::ArbitrumOne,
+            SupportedBlockchain::Base,
+            SupportedBlockchain::Gnosis,
+            SupportedBlockchain::BinanceSc,
+        ] {
+            if let Some(inquirer) = self.init_inquirer(blockchain).await {
+                let inquirer_clone = inquirer.clone();
+                tokio::spawn(async move {
+                    inquirer_clone.connect_to_all_nodes().await;
+                });
+            }
+        }
+    }
+
+    /// Initialize an inquirer for a specific blockchain
+    async fn init_inquirer(&self, blockchain: SupportedBlockchain) -> Option<Arc<EvmNodeInquirer>> {
+        let mut inquirers = self.inquirers.write().await;
+        if let std::collections::hash_map::Entry::Vacant(e) = inquirers.entry(blockchain) {
+            let inquirer = Arc::new(EvmNodeInquirer::new(blockchain, self.globaldb.clone()));
+            inquirer.connect_to_all_nodes().await;
+            e.insert(inquirer.clone());
+            Some(inquirer)
+        } else {
+            inquirers.get(&blockchain).cloned()
+        }
+    }
```

Comment:
but again this will connect to all the nodes in the list right? Why would you want this?

in python what we do is have the list of names and then iterate trough it, if the node is connected get it from the mapping and if not connect to it. Why did you decide to init all of them? doesn't it make sense to just init one and if it fails you init the next one? If there is some technical detail that I don't see that makes it dificult you can leave a TODO and we can see it in another PR but I'm not sure why you would want to init all of them.

Reply:
found a way around it. updated it.